File: iOS/Daily/Views/News/NewsArticleCard.swift
import SwiftUI
import DailyCore

/// Liquid Glass article card for feed lists featuring high-contrast typography, thumbnails, and quick actions.
public struct NewsArticleCard: View {
    public let article: NewsArticle
    public let onTap: () -> Void
    
    @ObservedObject private var savedService = SavedArticlesService.shared
    @ObservedObject private var settingsService = SettingsService.shared
    
    public init(article: NewsArticle, onTap: @escaping () -> Void) {
        self.article = article
        self.onTap = onTap
    }
    
    private var isReadLater: Bool {
        savedService.isReadLater(url: article.link)
    }
    
    private var isFavorite: Bool {
        savedService.isFavorite(url: article.link)
    }
    
    public var body: some View {
        Button {
            onTap()
        } label: {
            GlassCard(cornerRadius: 20, padding: 16) {
                VStack(alignment: .leading, spacing: 14) {
                    // Header: Publication, Category Pill & Timestamp
                    HStack(alignment: .center, spacing: 8) {
                        // Publication Favicon
                        if let iconUrl = article.publicationIconUrl, let url = URL(string: iconUrl) {
                            AsyncImage(url: url) { phase in
                                switch phase {
                                case .success(let image):
                                    image.resizable()
                                        .scaledToFill()
                                        .frame(width: 18, height: 18)
                                        .clipShape(Circle())
                                default:
                                    Image(systemName: "newspaper.fill")
                                        .font(.system(size: 11))
                                        .foregroundColor(ThemeColors.accentCyan)
                                }
                            }
                        }
                        
                        Text(article.publicationName ?? "News")
                            .font(.system(size: 12, weight: .bold, design: .rounded))
                            .foregroundColor(ThemeColors.accentCyan)
                            .lineLimit(1)
                        
                        if let category = article.category {
                            Text(category.displayName)
                                .font(.system(size: 10, weight: .semibold))
                                .foregroundColor(.white.opacity(0.8))
                                .padding(.horizontal, 7)
                                .padding(.vertical, 2)
                                .background(
                                    Capsule()
                                        .fill(Color.white.opacity(0.08))
                                        .overlay(Capsule().strokeBorder(Color.white.opacity(0.12), lineWidth: 1))
                                )
                        }
                        
                        Spacer()
                        
                        Text(article.relativeTimeFormatted)
                            .font(.system(size: 11, weight: .medium))
                            .foregroundColor(ThemeColors.fgMutedDark)
                    }
                    
                    // Body: Title, Description & Thumbnail
                    HStack(alignment: .top, spacing: 14) {
                        VStack(alignment: .leading, spacing: 6) {
                            Text(article.title)
                                .font(.system(size: 16, weight: .bold, design: .rounded))
                                .foregroundColor(.white)
                                .lineLimit(3)
                                .multilineTextAlignment(.leading)
                            
                            if let desc = article.description, !desc.isEmpty {
                                Text(desc)
                                    .font(.system(size: 13, weight: .regular))
                                    .foregroundColor(ThemeColors.fgMutedDark)
                                    .lineLimit(2)
                                    .multilineTextAlignment(.leading)
                            }
                        }
                        
                        if let imgStr = article.imageUrl, let url = URL(string: imgStr) {
                            AsyncImage(url: url) { phase in
                                switch phase {
                                case .success(let image):
                                    image.resizable()
                                        .scaledToFill()
                                        .frame(width: 88, height: 88)
                                        .clipShape(RoundedRectangle(cornerRadius: 14))
                                        .overlay(
                                            RoundedRectangle(cornerRadius: 14)
                                                .strokeBorder(Color.white.opacity(0.15), lineWidth: 1)
                                        )
                                        .shadow(color: Color.black.opacity(0.3), radius: 6, x: 0, y: 3)
                                default:
                                    EmptyView()
                                }
                            }
                        }
                    }
                    
                    // Actions Footer: Read Later, Favorite, Share
                    HStack(spacing: 16) {
                        if let author = article.author, !author.isEmpty {
                            HStack(spacing: 4) {
                                Image(systemName: "person.fill")
                                    .font(.system(size: 10))
                                Text(author)
                                    .font(.system(size: 11, weight: .medium))
                            }
                            .foregroundColor(ThemeColors.fgMutedDark)
                            .lineLimit(1)
                        }
                        
                        Spacer()
                        
                        // Read Later Toggle
                        Button {
                            withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                                savedService.toggleReadLater(for: article)
                            }
                            triggerHaptic()
                        } label: {
                            Image(systemName: isReadLater ? "bookmark.fill" : "bookmark")
                                .font(.system(size: 14, weight: .semibold))
                                .foregroundColor(isReadLater ? ThemeColors.accentCyan : ThemeColors.fgMutedDark)
                                .frame(width: 32, height: 32)
                                .background(
                                    Circle()
                                        .fill(isReadLater ? ThemeColors.accentCyan.opacity(0.15) : Color.white.opacity(0.06))
                                )
                        }
                        .buttonStyle(.plain)
                        
                        // Favorite Toggle
                        Button {
                            withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                                savedService.toggleFavorite(for: article)
                            }
                            triggerHaptic()
                        } label: {
                            Image(systemName: isFavorite ? "star.fill" : "star")
                                .font(.system(size: 14, weight: .semibold))
                                .foregroundColor(isFavorite ? Color(red: 1.0, green: 0.8, blue: 0.2) : ThemeColors.fgMutedDark)
                                .frame(width: 32, height: 32)
                                .background(
                                    Circle()
                                        .fill(isFavorite ? Color(red: 1.0, green: 0.8, blue: 0.2).opacity(0.15) : Color.white.opacity(0.06))
                                )
                        }
                        .buttonStyle(.plain)
                        
                        // Native Share Sheet
                        if let shareUrl = URL(string: article.link) {
                            ShareLink(item: shareUrl, subject: Text(article.title)) {
                                Image(systemName: "square.and.arrow.up")
                                    .font(.system(size: 13, weight: .semibold))
                                    .foregroundColor(ThemeColors.fgMutedDark)
                                    .frame(width: 32, height: 32)
                                    .background(
                                        Circle().fill(Color.white.opacity(0.06))
                                    )
                            }
                            .buttonStyle(.plain)
                        }
                    }
                }
            }
        }
        .buttonStyle(.plain)
    }
    
    private func triggerHaptic() {
        if settingsService.settings.hapticsEnabled {
            #if canImport(UIKit)
            UIImpactFeedbackGenerator(style: .light).impactOccurred()
            #endif
        }
    }
}

File: Android/app/src/main/java/com/intellidream/daily/presentation/news/NewsArticleCard.kt
package com.intellidream.daily.presentation.news

import android.content.Intent
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Bookmark
import androidx.compose.material.icons.rounded.BookmarkBorder
import androidx.compose.material.icons.rounded.CheckCircle
import androidx.compose.material.icons.rounded.Person
import androidx.compose.material.icons.rounded.Share
import androidx.compose.material.icons.rounded.Star
import androidx.compose.material.icons.rounded.StarBorder
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.intellidream.daily.designsystem.DailyAsyncImage
import com.intellidream.daily.designsystem.GlassCard
import com.intellidream.daily.designsystem.ThemeColors
import com.intellidream.daily.model.NewsArticle

@Composable
fun NewsArticleCard(
    article: NewsArticle,
    isReadLater: Boolean,
    isFavorite: Boolean,
    isMediumSubscribed: Boolean = false,
    onTap: () -> Unit,
    onToggleReadLater: () -> Unit,
    onToggleFavorite: () -> Unit,
    onFollowMediumAuthor: ((String) -> Unit)? = null
) {
    val context = LocalContext.current
    val haptic = LocalHapticFeedback.current

    GlassCard(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onTap() },
        cornerRadius = 20.dp,
        padding = 16.dp
    ) {
        Column(
            modifier = Modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            // Header: Publication, Category Pill & Relative Time
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                // Publication Favicon
                if (!article.publicationIconUrl.isNullOrBlank()) {
                    DailyAsyncImage(
                        url = article.publicationIconUrl,
                        contentDescription = null,
                        modifier = Modifier
                            .size(18.dp)
                            .clip(CircleShape),
                        contentScale = ContentScale.Crop
                    )
                }

                Text(
                    text = article.publicationName ?: "News",
                    color = ThemeColors.accentCyan,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis
                )

                val cat = article.category
                if (cat != null) {
                    Box(
                        modifier = Modifier
                            .clip(CircleShape)
                            .background(Color.White.copy(alpha = 0.08f))
                            .border(1.dp, Color.White.copy(alpha = 0.12f), CircleShape)
                            .padding(horizontal = 7.dp, vertical = 2.dp)
                    ) {
                        Text(
                            text = cat.displayName,
                            color = Color.White.copy(alpha = 0.8f),
                            fontSize = 10.sp,
                            fontWeight = FontWeight.SemiBold
                        )
                    }
                }

                Spacer(modifier = Modifier.weight(1f))

                Text(
                    text = article.relativeTimeFormatted,
                    color = ThemeColors.textMuted,
                    fontSize = 11.sp,
                    fontWeight = FontWeight.Medium
                )
            }

            // Body: Title, Description & Thumbnail
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(14.dp),
                verticalAlignment = Alignment.Top
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = article.title,
                        color = Color.White,
                        fontSize = 15.sp,
                        fontWeight = FontWeight.Bold,
                        maxLines = 3,
                        overflow = TextOverflow.Ellipsis,
                        lineHeight = 20.sp
                    )

                    val desc = article.description
                    if (!desc.isNullOrBlank()) {
                        Spacer(modifier = Modifier.height(6.dp))
                        Text(
                            text = desc,
                            color = ThemeColors.textSecondary,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.Normal,
                            maxLines = 2,
                            overflow = TextOverflow.Ellipsis,
                            lineHeight = 18.sp
                        )
                    }
                }

                if (!article.imageUrl.isNullOrBlank()) {
                    DailyAsyncImage(
                        url = article.imageUrl,
                        contentDescription = article.title,
                        modifier = Modifier
                            .size(88.dp)
                            .clip(RoundedCornerShape(14.dp))
                            .border(1.dp, Color.White.copy(alpha = 0.15f), RoundedCornerShape(14.dp)),
                        contentScale = ContentScale.Crop
                    )
                }
            }

            // Actions Footer: Author, Follow, Read Later, Favorite, Share
            Row(
                modifier = Modifier.fillMaxWidth(),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                val author = article.author
                if (!author.isNullOrBlank()) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                        modifier = Modifier.weight(1f, fill = false)
                    ) {
                        Icon(
                            imageVector = Icons.Rounded.Person,
                            contentDescription = null,
                            tint = ThemeColors.textMuted,
                            modifier = Modifier.size(12.dp)
                        )
                        Text(
                            text = author,
                            color = ThemeColors.textMuted,
                            fontSize = 11.sp,
                            fontWeight = FontWeight.Medium,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )

                        if (article.isMediumItem) {
                            val mediumUser = extractMediumUsername(article)
                            if (mediumUser != null) {
                                Spacer(modifier = Modifier.width(4.dp))
                                Box(
                                    modifier = Modifier
                                        .clip(CircleShape)
                                        .background(
                                            if (isMediumSubscribed) Color.White.copy(alpha = 0.06f)
                                            else ThemeColors.accentCyan.copy(alpha = 0.12f)
                                        )
                                        .clickable {
                                            haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                                            onFollowMediumAuthor?.invoke(mediumUser)
                                        }
                                        .padding(horizontal = 6.dp, vertical = 2.dp)
                                ) {
                                    Row(
                                        verticalAlignment = Alignment.CenterVertically,
                                        horizontalArrangement = Arrangement.spacedBy(3.dp)
                                    ) {
                                        Icon(
                                            imageVector = if (isMediumSubscribed) Icons.Rounded.CheckCircle else Icons.Rounded.Person,
                                            contentDescription = null,
                                            tint = if (isMediumSubscribed) Color.White.copy(alpha = 0.6f) else ThemeColors.accentCyan,
                                            modifier = Modifier.size(10.dp)
                                        )
                                        Text(
                                            text = if (isMediumSubscribed) "Following" else "Follow",
                                            color = if (isMediumSubscribed) Color.White.copy(alpha = 0.6f) else ThemeColors.accentCyan,
                                            fontSize = 10.sp,
                                            fontWeight = FontWeight.Bold
                                        )
                                    }
                                }
                            }
                        }
                    }
                }

                Spacer(modifier = Modifier.weight(1f))

                // Read Later Toggle
                IconButton(
                    onClick = {
                        haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                        onToggleReadLater()
                    },
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .background(
                            if (isReadLater) ThemeColors.accentCyan.copy(alpha = 0.18f)
                            else Color.White.copy(alpha = 0.06f)
                        )
                ) {
                    Icon(
                        imageVector = if (isReadLater) Icons.Rounded.Bookmark else Icons.Rounded.BookmarkBorder,
                        contentDescription = "Read Later",
                        tint = if (isReadLater) ThemeColors.accentCyan else ThemeColors.textMuted,
                        modifier = Modifier.size(16.dp)
                    )
                }

                // Favorite Toggle
                IconButton(
                    onClick = {
                        haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                        onToggleFavorite()
                    },
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .background(
                            if (isFavorite) Color(0xFFFFD700).copy(alpha = 0.18f)
                            else Color.White.copy(alpha = 0.06f)
                        )
                ) {
                    Icon(
                        imageVector = if (isFavorite) Icons.Rounded.Star else Icons.Rounded.StarBorder,
                        contentDescription = "Favorite",
                        tint = if (isFavorite) Color(0xFFFFD700) else ThemeColors.textMuted,
                        modifier = Modifier.size(16.dp)
                    )
                }

                // Share Button
                IconButton(
                    onClick = {
                        haptic.performHapticFeedback(HapticFeedbackType.LongPress)
                        val sendIntent = Intent().apply {
                            action = Intent.ACTION_SEND
                            putExtra(Intent.EXTRA_TEXT, "${article.title}\n\n${article.link}")
                            type = "text/plain"
                        }
                        context.startActivity(Intent.createChooser(sendIntent, "Share Article"))
                    },
                    modifier = Modifier
                        .size(32.dp)
                        .clip(CircleShape)
                        .background(Color.White.copy(alpha = 0.06f))
                ) {
                    Icon(
                        imageVector = Icons.Rounded.Share,
                        contentDescription = "Share",
                        tint = ThemeColors.textMuted,
                        modifier = Modifier.size(14.dp)
                    )
                }
            }
        }
    }
}

internal fun extractMediumUsername(article: NewsArticle): String? {
    if (!article.link.contains("medium.com")) return null
    return try {
        val uri = java.net.URI(article.link)
        val path = uri.path
        if (path != null && path.startsWith("/@")) {
            path.removePrefix("/@").substringBefore("/")
        } else if (uri.host?.contains(".medium.com") == true) {
            val sub = uri.host.substringBefore(".medium.com")
            if (sub != "www" && sub != "api") sub else null
        } else null
    } catch (_: Exception) {
        null
    }
}
